module/zfs/dbuf.c

```diff
@@ -1847,6 +1846,10 @@ dbuf_new_size(dmu_buf_impl_t *db, int size, dmu_tx_t *tx)
 	 * value and returning that up to our callers
 	 */
 	dmu_buf_will_dirty(&db->db, tx);
+	dr = list_head(&db->db_dirty_records);
```

Comment:
This should be moved back down so it happens while the `db->db_mtx` is held.  That said, it's not clear to me that coverity is actually wrong about this.  It's only because of how this code is currently structured and txgs are written that we know that after `dmu_buf_will_dirty()` is called there must be a valid dirty record.  This feels a bit dodgy to me.  How about this which at least explain why we believe this must be true.

```c
        /* dirty record added by dmu_buf_will_dirty() */
        dr = list_head(&db->db_dirty_records);
        VERIFY(dr != NULL);
```